### PR TITLE
[StandardToHandshake] Add loop handling networks

### PR DESCRIPTION
**Pull request by mortbopet**
While testing handshake accelerators, it has been a general issue that kernels were not able to be function pipelined. That is, having multiple activations of a kernel live, concurrently.
The reason for this is due to the previous use of control-merge operations at backedges in the CFG (loops). By doing so, multiple kernel invocations are able to enter the loop control network, which itself is already loop pipelining. In most scenarios, this then leads to a deadlock due to insufficient buffering inside the loop, due to having to accomodate multiple invocations + loop pipelining, all at once.

The solution to this is a new structure for loop entry control. Instead of a single control merge to allow for control entry to the loop, we now have a combination of a mux + a loop "priming" register. The priming register drives the mux, and the mux selects from either external control or internal, loop control. By doing so, we essentially guard loop control to be exclusively allowed for that inside of the loop, or an external loop activation. A similar mux structure is used for the inputs to the loop header; block data input muxes are split in two, containing inputs from external loop predecessors and internal loop predecessors, and a mux - driven by the loop priming register - selects between either of these, based on where control originated from.

image

The loop priming register is driven by the loop exit condition. Thus, when the loop exits, the priming mux is "re-primed" allowing for control to again enter from an external activation.

image

Related changes:
- Merge op "dataflow conversion check" has been relaxed to now not require that all merge-like ops have an operand from all predecessor blocks. This will not be the case for the externalCtrlMerge/loopCtrlMerge.
- Allow single-input muxes, but canonicalize them away. This simplifies lowering (removes a special case), and is more easily taken care of by a canonicalization pattern.

Marked as draft since:
- There are still some unsettled issues wrt. the legality of the in-loop pipelining being done through a control-merge. However, this PR changes nothing with regards to the existing method of loop pipelining - it only protects loops from function pipelining.
- i'll postpone reworking all of the tests until the PR has settled.

### CODE REVIEW

File: lib/Conversion/StandardToHandshake/StandardToHandshake.cpp [hunk 1]
```diff
@@ -804,20 +1206,6 @@ LogicalResult checkMergePredecessors(MergeLikeOpInterface mergeOp) {
              << getBlockPredecessorCount(block) << " predecessor blocks";
   }
 
-  // There must be a predecessor from each predecessor block
```
Comment (mikeurbach) [inline lib/Conversion/StandardToHandshake/StandardToHandshake.cpp:807]: I see this no longer holds for all merge-like ops, but are there cases where we should still enforce this on some ops?

**Review by mikeurbach — COMMENTED** [review]
I took a quick look and this looks good to me so far. I will take a deeper look at the lowerings later. Super exciting to see this functionality coming to Handshake.

File: lib/Conversion/StandardToHandshake/StandardToHandshake.cpp (hunk 1 above)
Comment (mortbopet) [reply lib/Conversion/StandardToHandshake/StandardToHandshake.cpp:807]: Hmm... This commit uses all three merge-like operations in a way where they will only have inputs from a subset of the predecessors. The check could be enabled if we carry in information about what we detected as loop headers (And therefore, rewrote) and exclude these from the check.

Comment (mikeurbach) [reply lib/Conversion/StandardToHandshake/StandardToHandshake.cpp:807]: I see. Well it seems like a pretty big relaxing to the verifier for anything that isn't a loop header. If it's not a huge pain, being able to keep this check and just exclude loop headers seems valuable.

File: lib/Dialect/Handshake/HandshakeOps.cpp [hunk 2]
```diff
@@ -311,11 +311,24 @@ struct EliminateSimpleMuxesPattern : mlir::OpRewritePattern<MuxOp> {
   }
 };
 
+struct EliminateUnaryMuxesPattern : OpRewritePattern<MuxOp> {
```
Comment (mikeurbach) [inline lib/Dialect/Handshake/HandshakeOps.cpp:314]: Feel free to add the canonicalization here whenever, it makes sense to me.

File: lib/Dialect/Handshake/HandshakeOps.cpp [hunk 3]
```diff
@@ -435,6 +445,16 @@ void ControlMergeOp::build(OpBuilder &builder, OperationState &result,
   sost::addAttributes(result, inputs, type);
 }
 
+void ControlMergeOp::build(OpBuilder &builder, OperationState &result,
```
Comment (mikeurbach) [inline lib/Dialect/Handshake/HandshakeOps.cpp:448]: Same for things like builders, they seem good to go.

File: lib/Conversion/StandardToHandshake/StandardToHandshake.cpp [hunk 4]
```diff
@@ -631,6 +644,421 @@ int getBranchCount(Value val, Block *block) {
...
+                       block->getPredecessors().end()) > 1;
+}
+
+// The BFS callback provides the current block as an argument and returns
+// whether search should halt.
+enum class BFSNextState { Halt, SkipSuccessors, Continue };
+using BFSCallback = llvm::function_ref<BFSNextState(Block *)>;
+
+static void blockBFS(Block *start, BFSCallback callback) {
+  std::set<Block *> visited;
```
Comment (mikeurbach) [inline lib/Conversion/StandardToHandshake/StandardToHandshake.cpp:658]: It might be worth switching to the LLVM data structures for this. Would a DenseSet and SmallVector work for `visited` and `queue`?

File: lib/Conversion/StandardToHandshake/StandardToHandshake.cpp [hunk 5]
```diff
@@ -631,6 +644,421 @@ int getBranchCount(Value val, Block *block) {
...
+  // loop latch, will then be selected between by a 3rd mux, based on the
+  // priming register.
+  for (MuxOp mux : muxesToReplace) {
+    auto externalDataMux = rewriter->create<MuxOp>(
+        loc, externalCtrlMerge.index(), externalDataInputs[mux]);
+
+    rewriter->replaceOp(
+        mux, rewriter
+                 ->create<MuxOp>(loc, primingRegister,
+                                 llvm::SmallVector<Value>{externalDataMux,
```
Comment (mikeurbach) [inline lib/Conversion/StandardToHandshake/StandardToHandshake.cpp:933]: nit: I think you could remove `llvm::SmallVector<Value>` here to save some typing.

File: lib/Conversion/StandardToHandshake/StandardToHandshake.cpp [hunk 6]
```diff
@@ -631,6 +644,421 @@ int getBranchCount(Value val, Block *block) {
...
+}
+
+LogicalResult LoopNetworkRewriter::processOuterLoop(Location loc,
+                                                    Block *loopHeader,
+                                                    Block *loopLatch) {
+  // To build the exit network, we need to determine the exit blocks of the
+  // loop, as well as all blocks contained within the loop. Exit blocks are the
+  // blocks that control is transfered to after exiting the loop. This is
+  // essentially determining the strongly connected components with the loop
+  // header. we perform a BFS from the loop header, and if the loop header is
```
Comment (mikeurbach) [inline lib/Conversion/StandardToHandshake/StandardToHandshake.cpp:999]: nit: could capitalize "We"

File: lib/Conversion/StandardToHandshake/StandardToHandshake.cpp [hunk 7]
```diff
@@ -631,6 +644,421 @@ int getBranchCount(Value val, Block *block) {
...
+    if (isReachable(currBlock, loopHeader)) {
+      inLoop.insert(currBlock);
+      return BFSNextState::Continue;
+    }
+    exitBlocks.insert(currBlock);
+    return BFSNextState::SkipSuccessors;
+  });
+
+  assert(inLoop.size() >= 2 && "A loop must have at least 2 blocks");
+  assert(exitBlocks.size() != 0 && "No exit blocks in the loop...?");
```
Comment (mikeurbach) [inline lib/Conversion/StandardToHandshake/StandardToHandshake.cpp:1012]: nit: maybe rephrase this like the other assertion? "A loop must have an exit block"

**Review by mikeurbach — APPROVED** [review]
I think this is a great feature, and the implementation looks good. One other thought: maybe start the feature defaulting to off in the initial commit, which would reduce the test churn and keep the previous behavior by default. It could be flipped (and all the tests updated) any time later.

File: lib/Conversion/StandardToHandshake/StandardToHandshake.cpp (hunk 5 above)
Comment (mortbopet) [reply lib/Conversion/StandardToHandshake/StandardToHandshake.cpp:933]: The `ValueRange` doesn't seem to be able to implicitly instantiate just using an initializer list.

**Comment by mortbopet** [comment]
@mikeurbach thank you for the review; I've re-enabled the verifier discussed earlier in cases where task pipelining is disabled. Regarding the tests, i think I'd prefer to keep task pipelining enabled by default (and thus the tests) since that seems to be what is currently used (by us, in CIRCT-HLS). I'll add a test for disabling task pipelining just to ensure that nothing gets lost in doing so.

File: lib/Conversion/StandardToHandshake/StandardToHandshake.cpp (hunk 5 above)
Comment (mikeurbach) [reply lib/Conversion/StandardToHandshake/StandardToHandshake.cpp:933]: Hmm gotcha I thought I'd done that before.
